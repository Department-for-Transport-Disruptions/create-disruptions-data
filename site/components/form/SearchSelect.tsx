import { Fragment, ReactElement, useState } from "react";
import Select, { SingleValue } from "react-select";
import AsyncSelect from "react-select/async";
import FormElementWrapper, { FormGroupWrapper } from "./FormElementWrapper";
import Table from "./Table";
import { ErrorInfo } from "../../interfaces";

interface SearchSelectProps<T> {
    placeholder?: string;
    inputName: string;
    initialErrors?: ErrorInfo[];
    getOptionLabel?: (value: T) => string;
    loadOptions: (inputValue: string) => Promise<T[]>;
    handleChange: (value: SingleValue<T>) => void;
    tableData: T[] | undefined;
    getRows: () => {
        header?: string;
        cells: (string | JSX.Element)[];
    }[];
    selected: SingleValue<T>;
    getOptionValue?: (value: T) => string;
    display: string;
    displaySize?: string;
    inputId: string;
    hint?: string;
    isClearable?: boolean;
    isAsync?: boolean;
    options?: T[] | undefined;
}
const SearchSelect = <T extends object>({
    selected,
    inputName,
    initialErrors = [],
    placeholder = "Select...",
    getOptionLabel,
    loadOptions,
    handleChange,
    tableData,
    getRows,
    getOptionValue,
    hint,
    inputId,
    display,
    displaySize = "s",
    isClearable = false,
    isAsync = true,
    options = [],
}: SearchSelectProps<T>): ReactElement => {
    const [searchInput, setSearchInput] = useState("");
    const handleInputChange = (value: string) => {
        setSearchInput(value);
    };

    return (
        <FormGroupWrapper errorIds={[inputId]} errors={initialErrors}>
            <div className="govuk-form-group">
                <label className={`govuk-label govuk-label--${displaySize}`} htmlFor={`${inputId}-input`}>
                    {display}
                </label>
                {hint ? (
                    <div id={`${inputId}-hint`} className="govuk-hint">
                        {hint}
                    </div>
                ) : null}
                <FormElementWrapper errors={initialErrors} errorId={inputId} errorClass="govuk-input--error">
                    {isAsync ? (
                        <AsyncSelect
                            isSearchable
                            styles={{
                                control: (baseStyles, state) => ({
                                    ...baseStyles,
                                    fontFamily: "GDS Transport, arial, sans-serif",
                                    border: "black solid 3px",
                                    outline: state.isFocused ? "#ffdd00 solid 3px" : "none",
                                    color: state.isFocused ? "white" : "black",
                                    marginBottom: "20px",
                                    "&:hover": { borderColor: "black" },
                                    width: "75%",
                                }),
                                option: (provided, state) => ({
                                    ...provided,
                                    color: state.isFocused ? "white" : "black",
                                    backgroundColor: state.isFocused ? "#3399ff" : "white",
                                }),
                            }}
                            cacheOptions
                            defaultOptions
                            value={selected}
                            placeholder={placeholder}
                            getOptionLabel={getOptionLabel}
                            getOptionValue={getOptionValue}
                            loadOptions={loadOptions}
                            onInputChange={handleInputChange}
                            inputValue={searchInput}
                            onChange={handleChange}
                            id={inputId}
                            instanceId={`async-dropdown-${inputName}`}
                            inputId={`async-dropdown-${inputName}-value`}
                            menuPlacement="auto"
                            menuPosition="fixed"
                            isClearable={isClearable}
                        />
                    ) : (
                        <Select
                            isSearchable
                            styles={{
                                control: (baseStyles, state) => ({
                                    ...baseStyles,
                                    fontFamily: "GDS Transport, arial, sans-serif",
                                    border: "black solid 3px",
                                    outline: state.isFocused ? "#ffdd00 solid 3px" : "none",
                                    color: state.isFocused ? "white" : "black",
                                    marginBottom: "20px",
                                    "&:hover": { borderColor: "black" },
                                    width: "75%",
                                }),
                                option: (provided, state) => ({
                                    ...provided,
                                    color: state.isFocused ? "white" : "black",
                                    backgroundColor: state.isFocused ? "#3399ff" : "white",
                                }),
                            }}
                            value={selected}
                            placeholder={placeholder}
                            getOptionLabel={getOptionLabel}
                            getOptionValue={getOptionValue}
                            options={options}
                            onInputChange={handleInputChange}
                            inputValue={searchInput}
                            onChange={handleChange}
                            id={inputId}
                            instanceId={`dropdown-${inputName}`}
                            inputId={`dropdown-${inputName}-value`}
                            menuPlacement="auto"
                            menuPosition="fixed"
                            isClearable={isClearable}
                        />
                    )}
                </FormElementWrapper>

                <Table rows={tableData ? getRows() : []} />
                {(tableData || []).map((element, index) => (
                    <Fragment key={`${inputName}-${index}`}>
                        <input type="hidden" name={`${inputName}${index + 1}`} value={JSON.stringify(element)} />
                    </Fragment>
                ))}
            </div>
        </FormGroupWrapper>
    );
};

export default SearchSelect;
