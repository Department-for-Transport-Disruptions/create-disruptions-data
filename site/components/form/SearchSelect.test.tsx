import renderer from "react-test-renderer";
import { describe, it, expect } from "vitest";
import SearchSelect from "./SearchSelect";

interface ColourOption {
    readonly value: string;
    readonly label: string;
}

describe("SearchSelect", () => {
    it("should render correctly with no errors", () => {
        const tree = renderer
            .create(
                <SearchSelect<ColourOption>
                    selected={{ value: "ocean", label: "Ocean" }}
                    inputName="test-id"
                    // eslint-disable-next-line @typescript-eslint/no-unused-vars
                    loadOptions={(_inputValue: string, _callback: (options: ColourOption[]) => void) =>
                        new Promise(() => [
                            { value: "blue", label: "Blue" },
                            { value: "purple", label: "Purple" },
                        ])
                    }
                    handleChange={() => null}
                    tableData={[]}
                    getRows={() => {
                        return [
                            {
                                header: "test row 1",
                                cells: ["1"],
                            },
                            {
                                header: "test row 2",
                                cells: ["2"],
                            },
                        ];
                    }}
                    display="Test Data"
                    displaySize="l"
                    inputId="test-id"
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render correctly with errors", () => {
        const tree = renderer
            .create(
                <SearchSelect<ColourOption>
                    selected={{ value: "ocean", label: "Ocean" }}
                    inputName="test-id"
                    // eslint-disable-next-line @typescript-eslint/no-unused-vars
                    loadOptions={(_inputValue: string, _callback: (options: ColourOption[]) => void) =>
                        new Promise(() => [
                            { value: "blue", label: "Blue" },
                            { value: "purple", label: "Purple" },
                        ])
                    }
                    handleChange={() => null}
                    tableData={[]}
                    getRows={() => {
                        return [
                            {
                                header: "test row 1",
                                cells: ["1"],
                            },
                            {
                                header: "test row 2",
                                cells: ["2"],
                            },
                        ];
                    }}
                    display="Test Data"
                    displaySize="l"
                    inputId="test-id"
                    initialErrors={[{ errorMessage: "There was an error", id: "test-id" }]}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
