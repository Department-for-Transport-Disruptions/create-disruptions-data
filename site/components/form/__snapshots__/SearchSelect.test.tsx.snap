// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchSelect > should render correctly with errors 1`] = `
<div
  className="govuk-form-group govuk-form-group--error"
>
  <div
    className="govuk-form-group"
  >
    <label
      className="govuk-label govuk-label--l"
      htmlFor="test-id-input"
    >
      Test Data
    </label>
    <div
      className=""
    >
      <span
        className="govuk-error-message"
        id="test-id-error"
      >
        <span
          className="govuk-visually-hidden"
        >
          Error: 
        </span>
        There was an error
      </span>
      <div
        className="govuk-input--error css-b62m3t-container"
        id="test-id"
        onKeyDown={[Function]}
      >
        <span
          className="css-1f43avz-a11yText-A11yText"
          id="react-select-dropdown-test-id-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          className="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          className=" css-1kxfjwu-control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className=" css-1fdsijx-ValueContainer"
          >
            <div
              className=" css-1dimb5e-singleValue"
            >
              Ocean
            </div>
            <div
              className=" css-qbdosj-Input"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-expanded={false}
                aria-haspopup={true}
                autoCapitalize="none"
                autoComplete="off"
                autoCorrect="off"
                className=""
                disabled={false}
                id="test-id-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                role="combobox"
                spellCheck="false"
                style={
                  {
                    "background": 0,
                    "border": 0,
                    "color": "inherit",
                    "font": "inherit",
                    "gridArea": "1 / 2",
                    "label": "input",
                    "margin": 0,
                    "minWidth": "2px",
                    "opacity": 1,
                    "outline": 0,
                    "padding": 0,
                    "width": "100%",
                  }
                }
                tabIndex={0}
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              className=" css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              className=" css-1xc3v61-indicatorContainer"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-tj5bde-Svg"
                focusable="false"
                height={20}
                viewBox="0 0 20 20"
                width={20}
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <table
      className="govuk-table"
    >
      <thead
        className="govuk-table__head"
      >
        <tr
          className="govuk-table__row"
        />
      </thead>
      <tbody
        className="govuk-table__body"
      >
        <tr
          className="govuk-table__row"
        >
          <th
            className="govuk-table__header align-middle"
            scope="row"
          >
            test row 1
          </th>
          <td
            className="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
          >
            1
          </td>
        </tr>
        <tr
          className="govuk-table__row"
        >
          <th
            className="govuk-table__header align-middle"
            scope="row"
          >
            test row 2
          </th>
          <td
            className="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
          >
            2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`SearchSelect > should render correctly with no errors 1`] = `
<div
  className="govuk-form-group"
>
  <div
    className="govuk-form-group"
  >
    <label
      className="govuk-label govuk-label--l"
      htmlFor="test-id-input"
    >
      Test Data
    </label>
    <div
      className=""
    >
      <div
        className=" css-b62m3t-container"
        id="test-id"
        onKeyDown={[Function]}
      >
        <span
          className="css-1f43avz-a11yText-A11yText"
          id="react-select-dropdown-test-id-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          className="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          className=" css-1kxfjwu-control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className=" css-1fdsijx-ValueContainer"
          >
            <div
              className=" css-1dimb5e-singleValue"
            >
              Ocean
            </div>
            <div
              className=" css-qbdosj-Input"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-expanded={false}
                aria-haspopup={true}
                autoCapitalize="none"
                autoComplete="off"
                autoCorrect="off"
                className=""
                disabled={false}
                id="test-id-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                role="combobox"
                spellCheck="false"
                style={
                  {
                    "background": 0,
                    "border": 0,
                    "color": "inherit",
                    "font": "inherit",
                    "gridArea": "1 / 2",
                    "label": "input",
                    "margin": 0,
                    "minWidth": "2px",
                    "opacity": 1,
                    "outline": 0,
                    "padding": 0,
                    "width": "100%",
                  }
                }
                tabIndex={0}
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              className=" css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              className=" css-1xc3v61-indicatorContainer"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-tj5bde-Svg"
                focusable="false"
                height={20}
                viewBox="0 0 20 20"
                width={20}
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <table
      className="govuk-table"
    >
      <thead
        className="govuk-table__head"
      >
        <tr
          className="govuk-table__row"
        />
      </thead>
      <tbody
        className="govuk-table__body"
      >
        <tr
          className="govuk-table__row"
        >
          <th
            className="govuk-table__header align-middle"
            scope="row"
          >
            test row 1
          </th>
          <td
            className="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
          >
            1
          </td>
        </tr>
        <tr
          className="govuk-table__row"
        >
          <th
            className="govuk-table__header align-middle"
            scope="row"
          >
            test row 2
          </th>
          <td
            className="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
          >
            2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
